fairscale/nn/data_parallel/fully_sharded_data_parallel.py

```diff
@@ -1534,12 +1543,14 @@ def alloc_storage_(data: torch.Tensor, size: torch.Size) -> None:
 
 
 def _post_state_dict_hook(
-    module: nn.Module, state_dict: "OrderedDict[str, torch.Tensor]", prefix: str, *args: Any
+    module: FullyShardedDataParallel, state_dict: "OrderedDict[str, torch.Tensor]", prefix: str, *args: Any
 ) -> "OrderedDict[str, torch.Tensor]":
-    if module.training_state == TrainingState.SUMMON_FULL_PARAMS:
-        # We copy the state_dict since full param will be freed after
-        # we exit the summon_full_params() context.
-        for key in state_dict.keys():
+    for key in state_dict.keys():
+        if state_dict[key].device.type != module.state_dict_device.type:
+            state_dict[key] = state_dict[key].to(device=module.state_dict_device)
+        elif module.training_state == TrainingState.SUMMON_FULL_PARAMS:
+            # We copy the state_dict since full param will be freed after
+            # we exit the summon_full_params() context.
             state_dict[key] = state_dict[key].clone()
```

Comment:
I am a bit hazy on when after calling model.state_dict(), is the state dict is moved to CPU?
What happens when the full state dict is summoned on rank 0 (for example), the accumulation of the full state dict still happens on the GPU associated with rank 0, right?

Reply:
This is a hook that's called after each FSDP instance's state_dict method is called. So if we have a nested FSDP wrapper around an expert layer, we'll call state_dict for that layer, which will summon the params on GPU, then this hook will copy it to CPU, and then finally the params should be freed.

This happens recursively for all FSDP instances, but there's actually a bug here (which I'm working on fixing), since this state_dict is actually built up incrementally so we have something like `n**2` calls to `clone()`, for `n` nested FSDP instances.

Reply:
> there's actually a bug here (which I'm working on fixing)

This is fixed now